forge-game/src/main/java/forge/game/ability/effects/DiscoverEffect.java

```diff
@@ -0,0 +1,169 @@
+package forge.game.ability.effects;
+
+import com.google.common.collect.Lists;
+import forge.card.CardStateName;
+import forge.game.Game;
+import forge.game.ability.AbilityKey;
+import forge.game.ability.AbilityUtils;
+import forge.game.ability.SpellAbilityEffect;
+import forge.game.card.*;
+import forge.game.cost.CostDiscard;
+import forge.game.cost.CostPart;
+import forge.game.cost.CostReveal;
+import forge.game.player.Player;
+import forge.game.player.PlayerController;
+import forge.game.spellability.LandAbility;
+import forge.game.spellability.SpellAbility;
+import forge.game.zone.PlayerZone;
+import forge.game.zone.Zone;
+import forge.game.zone.ZoneType;
+import forge.util.CardTranslation;
+import forge.util.Lang;
+import forge.util.Localizer;
+
+import java.util.List;
+import java.util.Map;
+
+public class DiscoverEffect extends SpellAbilityEffect {
+
+    @Override
+    protected String getStackDescription(SpellAbility sa) {
+        final StringBuilder sb = new StringBuilder();
+        List<Player> players = getTargetPlayers(sa);
+        sb.append(Lang.joinHomogenous(players)).append(" discover").append(players.size() > 1 ? " " : "s " );
+        sb.append(sa.getParamOrDefault("Num", "1")).append(".");
+
+        return sb.toString();
+    }
+
+    @Override
+    public void resolve(SpellAbility sa) {
+        final Card host = sa.getHostCard();
+        final Game game = host.getGame();
+
+        // Exile cards from the top of your library until you exile a nonland card with <N> mana value or less.
+        String type = "Card.nonLand+cmcLE" + (sa.getParamOrDefault("Num", "1"));
+
+        for (final Player p : getTargetPlayers(sa)) {
```

Comment:
target players is not needed, Discover only allows you to do it

Reply:
Sure, would be easy enough to add it back if/when some pushed card allows multiple players to do it